source/delegate/contracts/DelegateV2.sol

```diff
@@ -0,0 +1,398 @@
+  mapping(address => mapping(address => uint256)) public totalActiveRules;
+
+  // ERC-20 (fungible token) interface identifier (ERC-165)
+  bytes4 internal constant ERC20_INTERFACE_ID = 0x36372b07;
+
+  // The protocol identifier for setting intents on an Index
+  bytes2 public protocol;
+
+  /**
+   * @notice Contract Constructor
+   * @dev owner defaults to msg.sender if delegateContractOwner is provided as address(0)
+   * @param delegateSwap address Swap contract the delegate will deploy with
+   * @param delegateIndexer address Indexer contract the delegate will deploy with
+   * @param delegateContractOwner address Owner of the delegate
+   * @param delegateTradeWallet address Wallet the delegate will trade from
+   * @param delegateProtocol bytes2 The protocol identifier for Delegate contracts
+   */
+  constructor(
+    ISwap delegateSwap,
+    IIndexer delegateIndexer,
+    address delegateContractOwner,
+    address delegateTradeWallet,
+    bytes2 delegateProtocol
+  ) public {
+    swapContract = delegateSwap;
+    indexer = delegateIndexer;
+    protocol = delegateProtocol;
+
+    // If no delegate owner is provided, the deploying address is the owner.
+    if (delegateContractOwner != address(0)) {
+      transferOwnership(delegateContractOwner);
+    }
+
+    // If no trade wallet is provided, the owner's wallet is the trade wallet.
+    if (delegateTradeWallet != address(0)) {
+      tradeWallet = delegateTradeWallet;
+    } else {
+      tradeWallet = owner();
+    }
+
+    // Ensure that the indexer can pull funds from delegate account.
+    require(
+      IERC20(indexer.stakingToken()).approve(address(indexer), MAX_INT),
+      "STAKING_APPROVAL_FAILED"
+    );
+  }
+
+  function createRule(
+    address senderToken,
+    address signerToken,
+    uint256 senderAmount,
+    uint256 signerAmount
+  ) external {
+    // Rules created now holds this rule's ID
+    ruleIDCounter += 1;
+
+    // prev and next rule IDs are initialised to 0
+    rules[ruleIDCounter] = Rule({
+      signerToken: signerToken,
+      senderToken: senderToken,
+      senderAmount: senderAmount,
+      signerAmount: signerAmount,
+      nextRuleID: NO_RULE,
+      prevRuleID: NO_RULE
+    });
+
+    if (firstRuleID[senderToken][signerToken] == NO_RULE) {
+      firstRuleID[senderToken][signerToken] = ruleIDCounter;
+    } else {
+      _insertRuleInList(senderToken, signerToken, ruleIDCounter);
+    }
+
+    totalActiveRules[senderToken][signerToken] += 1;
+
+    emit CreateRule(
+      owner(),
+      ruleIDCounter,
+      senderToken,
+      signerToken,
+      senderAmount,
+      signerAmount
+    );
+  }
+
+  function deleteRule(uint256 ruleID) external {
+    _deleteRule(ruleID);
+  }
+
+  function provideOrder(Types.Order calldata order) external {
+    uint256 ruleID = firstRuleID[order.sender.token][order.signer.token];
+
+    require(order.signature.v != 0, "SIGNATURE_MUST_BE_SENT");
```

Comment:
Do we still need to require signature here?

Reply:
The security review of that was another point on the list of things to consider for V2, but I haven't done that analysis so didn't want to remove it. Happy to remove it once we've sat down and gone over all of the consequences of removing it